rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp

```diff
@@ -543,82 +543,105 @@ void SqliteStorage::prepare_for_writing()
+  std::string included_topic_names_str;
+  // Add topic names
   if (!storage_filter.topics.empty()) {
     for (auto & topic : storage_filter.topics) {
-      topic_and_service_list += "'" + topic + "'";
+      included_topic_names_str += "'" + topic + "'";
       if (&topic != &storage_filter.topics.back()) {
-        topic_and_service_list += ",";
+        included_topic_names_str += ",";
       }
     }
   }
 
-  // add service event topic name
+  // Add service event topic names
   if (!storage_filter.services_events.empty()) {
-    if (!topic_and_service_list.empty()) {
-      topic_and_service_list += ",";
+    if (!included_topic_names_str.empty()) {
+      included_topic_names_str += ",";
     }
     for (auto & service : storage_filter.services_events) {
-      topic_and_service_list += "'" + service + "'";
+      included_topic_names_str += "'" + service + "'";
       if (&service != &storage_filter.services_events.back()) {
-        topic_and_service_list += ",";
+        included_topic_names_str += ",";
       }
     }
   }
 
-  std::string topic_filter_str;
-  if (!topic_and_service_list.empty()) {
-    topic_filter_str.append("(topics.name IN (" + topic_and_service_list + "))");
+  std::string topics_filter_str;
+  if (!included_topic_names_str.empty()) {
+    topics_filter_str.append("(topics.name IN (" + included_topic_names_str + "))");
   }
 
   std::string regex_filter_str;
-  // add topic filter based on regular expression
+  // Add topics filter based on regular expression
   if (!storage_filter.regex.empty()) {
+    // Construct string for selected topics
     regex_filter_str = "(topics.name REGEXP '" + storage_filter.regex + "')";
   }
 
-  if (!topic_filter_str.empty() && !regex_filter_str.empty()) {
-    where_conditions.push_back("(" + topic_filter_str + " OR " + regex_filter_str + ")");
-  } else if (!topic_filter_str.empty()) {
-    where_conditions.push_back(topic_filter_str);
+  static const char * regex_for_all_service_events_str = "(topics.name REGEXP '.*/_service_event')";
+
+  if (!topics_filter_str.empty() && !regex_filter_str.empty()) {
+    // Note: Inclusive filter conditions shall be joined with OR
+    if (!storage_filter.services_events.empty()) {
+      where_conditions.push_back("(" + topics_filter_str + " OR " + regex_filter_str + ")");
+    } else {  // if services_events.empty we shall include all service events
+      where_conditions.push_back(
+        "(" + topics_filter_str + " OR " + regex_filter_str + " OR " +
+        regex_for_all_service_events_str + ")");
+    }
+  } else if (!topics_filter_str.empty()) {
+    if (!storage_filter.services_events.empty()) {
+      if (!storage_filter.topics.empty()) {
+        where_conditions.push_back(topics_filter_str);
+      } else {  // if topics list empty and service_events not empty we shall include all topics
+        where_conditions.push_back(
+          "(" + topics_filter_str + " OR NOT " + regex_for_all_service_events_str + ")");
+      }
```

Comment:
In this particular case is safe to include `all_topics` because we will get to this line only if `regex_filter_str` is empty.
 if (!topics_filter_str.empty() && !regex_filter_str.empty()) {
    where_conditions.push_back("(" + topics_filter_str + " OR " + regex_filter_str + ")");
  } else if (!topics_filter_str.empty()) {  // Note: regex_filter_str is empty at the same time
//   ...
  }
I will add a comment with a note.

Reply:
Thanks. I was wrong.